src/fold.rs

```diff
@@ -0,0 +1,984 @@
+        noop_fold_assoc_type_binding(self, type_binding)
+    }
+    fn fold_attribute(&mut self, _attr: Attribute) -> Attribute {
+        noop_fold_attribute(self, _attr)
+    }
+    fn fold_fn_ret_ty(&mut self, ret_ty: FunctionRetTy) -> FunctionRetTy {
+        noop_fold_fn_ret_ty(self, ret_ty)
+    }
+    fn fold_const_expr(&mut self, expr: ConstExpr) -> ConstExpr {
+        noop_fold_const_expr(self, expr)
+    }
+    fn fold_lit(&mut self, _lit: Lit) -> Lit {
+        noop_fold_lit(self, _lit)
+    }
+
+    fn fold_mac(&mut self, mac: Mac) -> Mac { panic!("fold_mac disabled by
+        default!\n Using a Folder on an AST before macro expansion is not
+        typically what you want, since a Folder renaming item names in a module
+        will miss all of those that are created by the expansion of a macro. If
+        you have an AST with unexpanded macros, first use the `MacroExpander`
+        Folder to expand all the macros, and then apply your Folder on the
+        resulting AST.\n If you really want to implement a folder that works on
+        ASTs containing unexpanded macros (for example if you are implementing
+        something like the `MacroExpander` folder), override the default
+        `Folder::fold_mac` implementation using, e.g., the
+        `folder::noop_fold_mac` function.");
+        // If you really want a folder that works on macros, use this definition
+        // in your trait impl to override the `fold_mac` method:
+        //
+        //noop_fold_mac(self, mac)
+    }
+
+    #[cfg(feature = "full")]
+    fn fold_crate(&mut self, _crate: Crate) -> Crate {
+        noop_fold_crate(self, _crate)
+    }
+    #[cfg(feature = "full")]
+    fn fold_item(&mut self, item: Item) -> Item {
+        noop_fold_item(self, item)
+    }
+    #[cfg(feature = "full")]
+    fn fold_expr(&mut self, expr: Expr) -> Expr {
+        noop_fold_expr(self, expr)
+    }
+    #[cfg(feature = "full")]
+    fn fold_foreign_item(&mut self, foreign_item: ForeignItem) -> ForeignItem {
+        noop_fold_foreign_item(self, foreign_item)
+    }
+    #[cfg(feature = "full")]
+    fn fold_pat(&mut self, pat: Pat) -> Pat {
+        noop_fold_pat(self, pat)
+    }
+    #[cfg(feature = "full")]
+    fn fold_fn_decl(&mut self, fn_decl: FnDecl) -> FnDecl {
+        noop_fold_fn_decl(self, fn_decl)
+    }
+    #[cfg(feature = "full")]
+    fn fold_trait_item(&mut self, trait_item: TraitItem) -> TraitItem {
+        noop_fold_trait_item(self, trait_item)
+    }
+    #[cfg(feature = "full")]
+    fn fold_impl_item(&mut self, impl_item: ImplItem) -> ImplItem {
+        noop_fold_impl_item(self, impl_item)
+    }
+    #[cfg(feature = "full")]
+    fn fold_method_sig(&mut self, method_sig: MethodSig) -> MethodSig {
+        noop_fold_method_sig(self, method_sig)
+    }
+    #[cfg(feature = "full")]
+    fn fold_stmt(&mut self, stmt: Stmt) -> Stmt {
+        noop_fold_stmt(self, stmt)
+    }
+
+    /// Prefer overriding fold_stmt instead!
```

Comment:
I think one often wants to fold over a `Block`, so `Folder` should offer a `.fold_block` method, but since a `Block` is just a vector of `Stmt`s, one typically wants to override `fold_stmt` instead (and not `fold_block`).